### PR TITLE
Re-enable group tests 1/?

### PR DESCRIPTION
**Pull request by raphaelrobert**
This PR re-enables the previously commented out group tests

### CODE REVIEW

File: tests/test_group.rs
Comment (franziskuskiefer) [inline tests/test_group.rs:10]: Tests should cover all supported ciphersuites.

File: tests/test_group.rs
Comment (franziskuskiefer) [inline tests/test_group.rs:68]: This more about the API. But having boolean arguments, especially in public APIs, is not great. It's never clear what the `true` here means. So for this API this should be something like `true /* force self-update */`. But maybe we should consider adding a separate function with this set to `true`, like `create_commit_self_update`.

File: tests/test_group.rs
Comment (franziskuskiefer) [inline tests/test_group.rs:86]: So we add Bob twice to the same group? Shouldn't this fail?

File: tests/test_group.rs
Comment (franziskuskiefer) [inline tests/test_group.rs:298]: This looks like a pretty crucial test. It should probably get an explanation.
Also, the `_print_tree` should only be interesting if this fails. But it won't be reached in the case the assert fails.

File: tests/test_group.rs
Comment (franziskuskiefer) [inline tests/test_group.rs:216]: What operations?
This needs a description of what's exactly happening/tested in here.

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
So this PR should enable the group tests. It's to expect that changes to the code are needed for this. But there's no explanation to which and why. So I only really looked at the test for now.
The description in the PR also doesn't appear to be correct as it only enables a part of the commented out test.

File: tests/test_group.rs
Comment (raphaelrobert) [reply tests/test_group.rs:86]: We only create a proposal/commit that we don't apply. This should not fail.

**Comment by raphaelrobert** [comment]
> So this PR should enable the group tests. It's to expect that changes to the code are needed for this. But there's no explanation to which and why. So I only really looked at the test for now.
> The description in the PR also doesn't appear to be correct as it only enables a part of the commented out test.

This PR indeed only covers a part of the tests. Enabling more of them would make the PR grow beyond the limit we imposed ourselves.
This PR fixes all the obvious problems that were introduced in the last weeks and that cause the inner workings to break.
There will be at least one follow-up PR that will close #77.

File: src/config.rs
Comment (franziskuskiefer) [inline src/config.rs:56]: Do we need these? The global config can be accessed already with the functions below.
Also, if we want these functions we should put them in the same `impl` block.

File: tests/test_group.rs
Comment (franziskuskiefer) [reply tests/test_group.rs:20]: 🤔

File: tests/test_group.rs
Comment (franziskuskiefer) [reply tests/test_group.rs:10]: blocked by #168

File: tests/test_group.rs
Comment (franziskuskiefer) [reply tests/test_group.rs:21]: 🤔

File: tests/test_group.rs
Comment (franziskuskiefer) [reply tests/test_group.rs:81]: Then there are comments missing explaining what this does (as mentioned in another comment already).

File: tests/test_group.rs
Comment (franziskuskiefer) [reply tests/test_group.rs:208]: 🤔

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
Maybe I'm missing something but I don't find a description of the changes.

File: tests/test_group.rs
Comment (raphaelrobert) [reply tests/test_group.rs:21]: This is something for #164 then, we don't want to do that here.

File: tests/test_group.rs
Comment (raphaelrobert) [reply tests/test_group.rs:208]: That was added, but I think you are looking at the wrong file

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
I still can't find a description of the changes.

**Comment by raphaelrobert** [comment]
> I still can't find a description of the changes.

The last commit only addresses your comments, as the name suggests. There are no other changes.

**Comment by franziskuskiefer** [comment]
> > I still can't find a description of the changes.
> 
> The last commit only addresses your comments, as the name suggests. There are no other changes.

There is no description of any changes in this PR. That's what I'm waiting for before reviewing. The description is "This PR re-enables the previously commented out group tests". But there are a ton of code changes in this PR. They all need a high level description here.

**Comment by raphaelrobert** [comment]
> > > I still can't find a description of the changes.
> > 
> > 
> > The last commit only addresses your comments, as the name suggests. There are no other changes.
> 
> There is no description of any changes in this PR. That's what I'm waiting for before reviewing. The description is "This PR re-enables the previously commented out group tests". But there are a ton of code changes in this PR. They all need a high level description here.

The context of your question wasn't clear.
There is a high-level description which you didn't quote, but here is a more detailed one:
 - Re-enable part of the group tests to test basic operations like Add & Update
 - Validate public keys from the UpdatePath
 - Change return type of KeyPackage verification from boolean to more specific errors
 - Fix direct path calculation in treemath
 - Fix missing RatchetTree clone in create_commit
 - Refactor part of PrivateTree
 - Fix various smaller issues that made the tests break

File: src/tree/private_tree.rs
Comment (franziskuskiefer) [inline src/tree/private_tree.rs:137]: This appears to be wrong. According to spec we have `path_secret[0] = DeriveSecret(leaf_secret, "path")`. So unless you changed the way this function is called the `start_secret` shouldn't be used directly.

File: src/tree/treemath.rs
Comment (franziskuskiefer) [inline src/tree/treemath.rs:139]: Please add a test for this.

File: src/tree/treemath.rs
Comment (franziskuskiefer) [inline src/tree/treemath.rs:159]: And this one as well.

File: tests/test_group.rs
Comment (franziskuskiefer) [reply tests/test_group.rs:10]: #168 got merged so we should add tests for all ciphersuites.

File: tests/test_group.rs
Comment (franziskuskiefer) [inline tests/test_group.rs:71]: Can we test more here than just the fact that there is a path?

File: tests/test_group.rs
Comment (franziskuskiefer) [inline tests/test_group.rs:74]: "... only Add Proposals this does not generate ..."

File: tests/test_group.rs
Comment (franziskuskiefer) [inline tests/test_group.rs:97]: Which one? Looks like the second without forced self-update.

File: tests/test_group.rs
Comment (franziskuskiefer) [inline tests/test_group.rs:106]: Shouldn't this `unwrap` be a test case earlier?

File: tests/test_group.rs
Comment (franziskuskiefer) [inline tests/test_group.rs:132]: `/* force self update */`

File: tests/test_group.rs
Comment (franziskuskiefer) [inline tests/test_group.rs:155]: Maybe use `.expect()`? There are a lot more like this in this file.

File: tests/test_group.rs
Comment (franziskuskiefer) [inline tests/test_group.rs:207]: To make this more digestible having a high level description of what's supposed to happen would be nice. Something like
* Alice creates group
* Alice adds Bob
* Bob removes Alice 
...

File: tests/test_group.rs
Comment (franziskuskiefer) [inline tests/test_group.rs:263]: Shouldn't the `create_commit_optional_path` test fail then?

File: tests/test_group.rs
Comment (franziskuskiefer) [inline tests/test_group.rs:270]: Adding a little more information to panics would help debugging later.

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
Thanks for the update and additional description. See inline comments

Please also fix the clippy warning.

Other changes in here (that should co into the commit message in the end):
*  new public function `as_application_message` to get a `MLSPlaintext` as application message (byte slice).

File: src/group/mls_group/apply_commit.rs
Comment (raphaelrobert) [reply src/group/mls_group/apply_commit.rs:86]: https://github.com/raphaelrobert/maelstrom/issues/171

File: src/group/mls_group/mod.rs
Comment (raphaelrobert) [reply src/group/mls_group/mod.rs:224]: The problem was that it was off by a factor of 2. The new indexing wasn't available when I wrote this, I'm adapting it now.

File: tests/test_group.rs
Comment (raphaelrobert) [reply tests/test_group.rs:132]: When you use rust-analyzer, you get that for free, that's why I don't make these kind of annotations.

File: tests/test_group.rs
Comment (raphaelrobert) [reply tests/test_group.rs:207]: Good point. I'll add this at the end in the last PR, because the order is still very much in flux. These tests need a better system anyway, to be able to cover more cases. For now the idea is to just re-enable the tests that were already written.

File: tests/test_group.rs
Comment (raphaelrobert) [reply tests/test_group.rs:263]: That comment was outdated

File: tests/test_group.rs
Comment (franziskuskiefer) [reply tests/test_group.rs:206]: Yeah, That's what @kkohbrok is currently working on (no PR yet).

File: src/key_packages/mod.rs
Comment (raphaelrobert) [reply src/key_packages/mod.rs:90]: Agreed. The problem right now is that the public error code all have integer values which prevents the nesting. We should probably revert on that decision for now and do this only for FFI bindings.

File: src/tree/mod.rs
Comment (raphaelrobert) [reply src/tree/mod.rs:52]: Yes, we have to make a complete copy of the tree, including the of the secrets. There might be a way around it, but it would require a crazy setup of references, and at this point I want to keep it simple.

File: src/tree/mod.rs
Comment (raphaelrobert) [reply src/tree/mod.rs:332]: No. the `debug_assert` didn't seem to make much sense, not sure how it ended up there.

File: src/tree/private_tree.rs
Comment (raphaelrobert) [reply src/tree/private_tree.rs:137]: You are right and that's described here: https://github.com/raphaelrobert/maelstrom/issues/165
It's already fixed in the follow-up PR.

File: src/tree/treemath.rs
Comment (raphaelrobert) [reply src/tree/treemath.rs:139]: This can only be tested with interop test vectors, recalculating the same thing is a moot point. We should do that during the IETF hackathon.

File: tests/test_group.rs
Comment (raphaelrobert) [reply tests/test_group.rs:71]: The follow-up PR does test more. We need the API change for that.

File: tests/test_group.rs
Comment (raphaelrobert) [reply tests/test_group.rs:106]: Do you mean we should test if the welcome message exists? I can add that test case.

**Comment by raphaelrobert** [comment]
Thanks for all the feedback @franziskuskiefer!

I addressed it in the last two commits, or through comments.

Unfortunately P256 is still broken so it didn't make it in here. I need to investigate what it is this time.

File: tests/test_group.rs
Comment (franziskuskiefer) [inline tests/test_group.rs:275]: "function"

File: tests/test_group.rs
Comment (franziskuskiefer) [reply tests/test_group.rs:97]: 🤔

File: tests/test_group.rs
Comment (franziskuskiefer) [reply tests/test_group.rs:81]: 🤔

File: tests/test_group.rs
Comment (franziskuskiefer) [reply tests/test_group.rs:106]: Yeah, we shouldn't just panic here when it's missing but check for this earlier.

File: src/tree/treemath.rs
Comment (franziskuskiefer) [reply src/tree/treemath.rs:139]: A simple test with predefined tree is certainly possible. That's all that's needed for this for now.
Of course we need more and better tests. But this code change should have a test. Same below.

File: src/tree/mod.rs
Comment (franziskuskiefer) [reply src/tree/mod.rs:52]: Any change like this will cost us a couple weeks to fix down the road. If you think that's worth it, you can keep it as is.

File: src/tree/mod.rs
Comment (franziskuskiefer) [reply src/tree/mod.rs:332]: Then we need a test 🙂

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
Almost there.

There are always a lot of unresolved and unanswered comments. This costs a lot of time to go through every time and increases the number of roundtrips we need to get this reviewed.

File: tests/test_group.rs
Comment (raphaelrobert) [reply tests/test_group.rs:106]: The check is now in L276

File: src/tree/treemath.rs
Comment (raphaelrobert) [reply src/tree/treemath.rs:139]: There's a test now in `treemath_test.rs`

File: src/tree/mod.rs
Comment (raphaelrobert) [reply src/tree/mod.rs:332]: I thought about this some more and concluded that this cannot occur with malformed payload. It must be a library error on the treemath level and I changed the check accordingly.

File: src/tree/mod.rs
Comment (raphaelrobert) [reply src/tree/mod.rs:52]: We can discuss this offline, but I don't see a big issue with copying the tree right now. This is also what the spec is suggesting.

**Review by franziskuskiefer — APPROVED** [review]
Ok lets get this in and address everything that's still open later.
Please make sure to squash before merging and make the commit message cover all changes in here.

**Pull request merged**
